syntaxes/src/expression/grammar.ts

```diff
@@ -0,0 +1,886 @@
+        '1': {
+          name: 'punctuation.definition.string.end.ts',
+        },
+        '2': {
+          name: 'invalid.illegal.newline.ts',
+        },
+      },
+      patterns: [
+        {
+          include: '#stringCharacterEscape',
+        },
+      ],
+    },
+
+    stringCharacterEscape: {
+      name: 'constant.character.escape.ts',
+      match: /\\(x\h{2}|[0-2][0-7]{0,2}|3[0-6][0-7]?|37[0-7]?|[4-7][0-7]?|.|$)/,
+    },
+
+    ternaryExpression: {
+      begin: /(?=\?)(?!\?\.)/,
+      end: /(?=$|"|[;,})\]])/,
+      patterns: [
+        {
+          include: '#ternaryOperator',
+        },
+        {
+          include: '#ngExpression',
+        },
+      ],
+    },
+
+    ternaryOperator: {
+      begin: /(\?)(?!\.)/,
+      beginCaptures: {
+        '0': {
+          name: 'keyword.operator.ternary.ts',
+        },
+      },
+      end: /(:)/,
+      endCaptures: {
+        '0': {
+          name: 'keyword.operator.ternary.ts',
+        },
+      },
+      patterns: [
+        {
+          include: '#ngExpression',
+        },
+      ],
+    },
+
+    thisLiteral: {
+      name: 'variable.language.this.ts',
+      match: /(?<!\.|\$)\bthis\b(?!\$)/,
+    },
+
+    tsString: {
+      patterns: [
+        {
+          include: '#qstringSingle',
+        },
+        {
+          include: '#qstringDouble',
+        },
+      ],
+    },
+
+    typeAnnotation: {
+      name: 'meta.type.annotation.ts',
+      begin: /:/,
+      beginCaptures: {
+        '0': {
+          name: 'keyword.operator.type.annotation.ts',
+        },
+      },
+      end: /(?=$|[,);\}\]]|\/\/|")|(?==[^>])|(?<=[\}>\]\)]|[_$[:alpha:]])\s*(?=\{)/,
+      patterns: [
+        {
+          include: '#type',
+        },
+      ],
+    },
+
+    typeBuiltinLiterals: {
+      name: 'support.type.builtin.ts',
+      match: /(?<!\.|\$)\b(this|true|false|undefined|null)\b(?!\$)/,
+    },
+
+    typeFnTypeParameters: {
+      patterns: [
+        {
+          name: 'meta.type.constructor.ts',
+          match: /(?<!\.|\$)\b(new)\b(?=\s*\<)/,
+          captures: {
+            '1': {
+              name: 'keyword.control.new.ts',
+            },
+          },
+        },
+        {
+          name: 'meta.type.constructor.ts',
+          begin: /(?<!\.|\$)\b(new)\b\s*(?=\()/,
+          beginCaptures: {
+            '1': {
+              name: 'keyword.control.new.ts',
+            },
+          },
+          end: /(?<=\))/,
+          patterns: [
+            {
+              include: '#functionParameters',
+            },
+          ],
+        },
+        {
+          name: 'meta.type.function.ts',
+          begin: /(?<=\>)\s*(?=\()/,
+          end: /(?<=\))/,
+          patterns: [
+            {
+              include: '#functionParameters',
+            },
+          ],
+        },
+        {
+          name: 'meta.type.function.ts',
+          begin: '(?x)( (?= [(]\\s*( ([)]) |  (\\.\\.\\.) | ([_$[:alnum:]]+\\s*( ([:,?=])| ([)]\\s*=>) )) ) ) )',
```

Comment:
Same as some previous comment. Are the spaces necessary? The string is okay; we can fix that up later.

Reply:
Most of this grammar I have it derived from [this file](https://github.com/microsoft/vscode-textmate/blob/master/test-cases/themes/syntaxes/TypeScript.tmLanguage.json#L1805) from the microsoft vscode-textmate repository. I don't think the spaces will make any difference to be honest. I'll remove it, and we can later restore it if needed

Reply:
I think these patterns beginning with the ignore-whitespace flag `(?x)` are a holdover from when they were converted from the XML format (in that format, patterns with this flag [can be split onto multiple lines](https://github.com/Microsoft/TypeScript-TmLanguage/blob/master/TypeScript.tmLanguage) which makes them a little more legible).

In my experience, the way to successfully refactor these is to remove the `(?x)` flag, then remove all whitespace characters (including `\n`s) and comments (beginning with `#` and usually ending with a bunch of whitespace characters) from the pattern.

From there, refactoring them to regex literal should just be a matter of replacing the demarcating quotes with `/` and removing the redundant escaping backslashes. We could also replace the Oniguruma POSIX brackets with their equivalent character classes for consistency and better TypeScript syntax highlighting (e.g. `[_$[:alnum:]]` becomes `[_$a-zA-Z0-9]` -- more info [here](https://github.com/stedolan/jq/wiki/Docs-for-Oniguruma-Regular-Expressions-(RE.txt))), but that's not super important IMO.

Reply:
I agree with replacing the Oniguruma POXIX brackets as they are not that intuitive. I think that we can hold that a bit because there are a few more of these

Reply:
I removed the spaces here too.